### DISCUSSION TITLE
Running Peridigm correctly on an HPC cluster

### DISCUSSION THREAD

**Discussion by jritta** [peridigm/peridigm#162] [Q&A] [answered]
Hello community,

How do you lunch your Peridigm application on a cluster?
With cluster I mean an HPC infrastructure that has several compute nodes which are connected by some fast network, e.g. InfiniBand. Each node has multiple physical CPUs, e.g. two CPUs with each 10 cores + SMT.

Lets assume following configuration: We use 2 nodes with each 2 CPUs which each has 10 cores + SMT. So we have 2x2x10 = 40 physical cores or with using SMT 40x2 = 80 threads: Do you lunch Peridigm simply by:
* `mpirun -np 80 Peridigm InputDeck.yaml` or 
* `mpirun -np 40 Peridigm InputDeck.yaml` or 
* something more sophisticated (`--bind-to core`,...)?

Thanks in advance!
Best,
Jonas

**Comment by djlittl** [comment, marked as answer]
Jonas,

The answer to your question is platform specific.  Most institutional clusters use a job submission system, such as SLURM, and the sys admin team that runs the cluster should have suggestions on the optimal settings.  It is through the job management system and/or operating system that mpi ranks are mapped to physical resources, Peridigm itself does not control that.

For your specific question about 40 versus 80 mpi ranks, my guess is that running with 80 mpi ranks will be significantly faster than 40 for the computing platform you describe.  But I doubt it will be twice as fast.  Truth is, it’s pretty hard to guess how different systems will perform under different conditions, so you’ll want to experiment to find your best run time options.  Sometimes even the choice of compiler or compiler flags can make a significant different.  You’ll want to read whatever user documentation is available for your system, particularly if It’s a large institutional system with dedicated user support.

Final quick comment:  there is nothing special about Peridigm that would make it different from any other mpi-based code running on a cluster.

Explicit dynamics will scale really well, and implicit (quasi-statics) will scale less well because the global linear solve is difficult to scale.

Hope this helps.

Dave

From: Jonas ***@***.***>
Reply-To: peridigm/peridigm ***@***.***>
Date: Friday, October 1, 2021 at 3:39 AM
To: peridigm/peridigm ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EXTERNAL] [peridigm/peridigm] Running Peridigm correctly on an HPC cluster (Discussion #162)


Hello community,

How do you lunch your Peridigm application on a cluster?
With cluster I mean an HPC infrastructure that has several compute nodes which are connected by some fast network, e.g. InfiniBand. Each node has multiple physical CPUs, e.g. two CPUs with each 10 cores + SMT.

Lets assume following configuration: We use 2 nodes with each 2 CPUs which each has 10 cores + SMT. So we have 2x2x10 = 40 physical cores or with using SMT 40x2 = 80 threads: Do you lunch Peridigm simply by:

  *   mpirun -np 80 Peridigm InputDeck.yaml or
  *   mpirun -np 40 Peridigm InputDeck.yaml or
  *   something more sophisticated (--bind-to core,...)?

Thanks in advance!
Best,
Jonas

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/peridigm/peridigm/discussions/162>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABKXABPIL6PATLP2IU3UQWLUEV6URANCNFSM5FEIY67Q>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Reply by jritta** [reply]
Hi Dave,

Thanks for your advice and estimation!
Knowing that in the case described above running with 80 MPI processes instead of 40 MPI processes will likely perform better helps me already. I will also try to contact the sys admins to get more insights on the specific system that is available to me.

One last question: Is there some advantage when I compile Trilinos with OpenMP? So does in this case Peridigm or more precisely Trilinos make use of some hybrid-parallelization (mix MPI  with OpenMP)?

Best,
Jonas
